[Navigation] Dispatch NavigateEvent on reload

rwlbuis opened the pull request:
#### 76a11dfefffc787445970c5150db31e82cca32de
<pre>
[Navigation] Dispatch NavigateEvent on reload
<a href="https://bugs.webkit.org/show_bug.cgi?id=273688">https://bugs.webkit.org/show_bug.cgi?id=273688</a>

Reviewed by Alex Christensen.

Various kinds of reloads, such as navigation.reload() and location.reload() need to dispatch the NavigateEvent [1]. Since a reload triggered
by browser UI (i.e. clicking the reload button on the crome UI) should not dispatch the NavigateEvent, put the logic in FrameLoader::loadFrameRequest
and not in FrameLoader::reload. And for the same reason, make Navigation::reload go through FrameLoader::changeLocation and thus FrameLoader::loadFrameRequest.

This PR also introduces logic to deal with a reload causing an existing window, and matching navigation, to be discarded in favour of a new window and navigation
combination, in that case we have to copy over the navigation entries from the previous Navigation instance.

[1] <a href="https://html.spec.whatwg.org/multipage/browsing-the-web.html#reloading-and-traversing">https://html.spec.whatwg.org/multipage/browsing-the-web.html#reloading-and-traversing</a>:fire-a-push/replace/reload-navigate-event

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/currententrychange-event/navigation-reload-intercept-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/navigate-destination-getState-reload-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-history-entry/key-id-location-reload-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/reload-info-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/reload-no-args-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/reload-state-and-info-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/reload-state-undefined-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/reload-detach-in-onnavigate-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/reload-intercept-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/reload-intercept-rejected-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/reload-preventDefault-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/reload-rejection-order-unload-unserializablestate-expected.txt: Removed.
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/per-entry-events/dispose-same-document-reload-with-intercept-expected.txt:
* Source/WebCore/loader/DocumentWriter.cpp:
(WebCore::DocumentWriter::begin):
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::didBeginDocument):
(WebCore::FrameLoader::loadFrameRequest):
(WebCore::FrameLoader::loadURL):
(WebCore::FrameLoader::loadPostRequest):
(WebCore::FrameLoader::dispatchNavigateEvent):
(WebCore::FrameLoader::updateNavigationAPIEntries): Deleted.
* Source/WebCore/loader/FrameLoader.h:
* Source/WebCore/page/Navigation.cpp:
(WebCore::Navigation::initializeForNewWindow):
(WebCore::Navigation::reload):
(WebCore::Navigation::notifyCommittedToEntry):
(WebCore::Navigation::abortOngoingNavigation):
(WebCore::Navigation::innerDispatchNavigateEvent):
(WebCore::Navigation::initializeEntries): Deleted.
* Source/WebCore/page/Navigation.h:
* Source/WebCore/page/NavigationHistoryEntry.cpp:
(WebCore::NavigationHistoryEntry::NavigationHistoryEntry):
(WebCore::NavigationHistoryEntry::create):
(WebCore::NavigationHistoryEntry::getState const):
(WebCore::NavigationHistoryEntry::setState):
* Source/WebCore/page/NavigationHistoryEntry.h:

Canonical link: <a href="https://commits.webkit.org/283604@main">https://commits.webkit.org/283604@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3a58fd6ec54ce7ac59c6ca9bc60b798e23741cab

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/66684 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/46059 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/19306 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/70718 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/17817 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/68802 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/53858 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/17583 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/53435 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/12016 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/69751 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/42406 "14 new passes 2 flakes 6 failures") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/57708 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/34103 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/39077 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/15099 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/16171 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/60992 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/15440 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/72420 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/10641 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/14800 "Found 1 new test failure: http/tests/IndexedDB/storage-limit-1.https.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/60761 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/10673 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/57764 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/61096 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14758 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/8762 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/2379 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/41866 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/42943 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/44126 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/42686 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebCore/loader/FrameLoader.cpp:
```diff
@@ -1428,6 +1428,15 @@ void FrameLoader::loadFrameRequest(FrameLoadRequest&& request, Event* event, Ref
+    if (loadType == FrameLoadType::Reload) {
+        if (m_frame->document() && m_frame->document()->settings().navigationAPIEnabled()) {
+            if (RefPtr domWindow = m_frame->document()->domWindow()) {
+                if (!domWindow->protectedNavigation()->dispatchPushReplaceReloadNavigateEvent(url, NavigationNavigationType::Reload, false))
```
Can this cause JS re-entry (JS calling C++ which calls JS on the same stack)?  I think it can, in which case this event should be scheduled on the event loop instead of called synchronously.

rwlbuis replied:
What kind of scenario are you thinking of? Are you worried about recursion/endless loop? My latest patch protects against detachment in the event handler btw.

rwlbuis replied:
I added the event loop scheduling, PTAL.

achristensen07 commented on LayoutTests/TestExpectations:
```diff
@@ -6938,7 +6938,8 @@ imported/w3c/web-platform-tests/navigation-api/scroll-behavior/scroll-on-synthet
 imported/w3c/web-platform-tests/navigation-api/updateCurrentEntry-method/cross-document-location-api.html [ Skip ]
-imported/w3c/web-platform-tests/navigation-api/updateCurrentEntry-method/location-reload.html [ Pass ]
+imported/w3c/web-platform-tests/navigation-api/navigate-event/navigate-meta-refresh.html [ Pass ]
+imported/w3c/web-platform-tests/navigation-api/navigate-event/navigate-destination-getState-reload.html [ Pass ]
```
I really think we should merge https://github.com/WebKit/WebKit/pull/27818 first.

achristensen07 commented on Source/WebCore/loader/FrameLoader.cpp:
```diff
@@ -1442,10 +1442,29 @@ void FrameLoader::loadFrameRequest(FrameLoadRequest&& request, Event* event, Ref
-        loadPostRequest(WTFMove(request), referrer, loadType, event, WTFMove(formState), WTFMove(completionHandler));
-    else
-        loadURL(WTFMove(request), referrer, loadType, event, WTFMove(formState), WTFMove(privateClickMeasurement), WTFMove(completionHandler));
+    auto finishLoadFrameRequest = [this, referrer, event, loadType, frame, completionHandler] (FrameLoadRequest&& request, RefPtr<FormState>&& formState, std::optional<PrivateClickMeasurement>&& privateClickMeasurement) mutable {
```
Don't capture this without a something managing the lifetime of this.  Here, I think you should capture a Ref<LocalFrame> and call LocalFrame::loader

rwlbuis replied:
Done, however it turns out due to recent changes we need stuff like https://bugs.webkit.org/show_bug.cgi?id=274170 in place to make this show up green. Patrick is working on that, but for now we can hold off action on this PR and also no need to run the tests on the bots yet.

TingPing replied:
#28576

achristensen07 commented on Source/WebCore/page/NavigationHistoryEntry.cpp:
```diff
@@ -40,14 +40,24 @@ namespace WebCore {
+    RefPtr state = historyItem->navigationAPIStateObject();
+    auto ret = adoptRef(*new NavigationHistoryEntry(context, WTFMove(historyItem), other.m_urlString, other.m_key, other.m_id));
+    ret->setState(WTFMove(state));
+    return ret;
```
It is typically an anti-pattern to call a setter immediately after a constructor.  It usually indicates that we should instead add a parameter to the constructor.  Why can't we just pass the context and other directly to the constructor and let the constructor handle it?

rwlbuis replied:
I did not pay much attention to this since I think we'll have to improve the state code further to deal with special cases. Nevertheless, even though this may not be the final code to deal with state, better to at least have some decent code in, fixed.

achristensen07 commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -809,7 +867,10 @@ bool Navigation::innerDispatchNavigateEvent(NavigationNavigationType navigationT
 
-        if (navigationType == NavigationNavigationType::Push || navigationType == NavigationNavigationType::Replace) {
+        if (navigationType == NavigationNavigationType::Reload) {
+            // Not in specification but matches chromium implementation and tests.
```
Should we raise a spec issue?

rwlbuis replied:
I filed https://github.com/whatwg/html/issues/10621.

achristensen07 commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -104,18 +105,63 @@ static std::optional<size_t> getEntryIndexOfHistoryItem(const Vector<Ref<Navigat
+
+            if (!SecurityOrigin::create(item->url())->isSameOriginAs(startingOrigin))
+                break;
+            items.insert(0, WTFMove(item));
```
Repeated prepending to a Vector is O(n^2) that can easily be changed to O(n).

rwlbuis replied:
Good point, should be fixed.

achristensen07 commented on Source/WebCore/loader/FrameLoader.cpp:
```diff
@@ -1440,10 +1441,30 @@ void FrameLoader::loadFrameRequest(FrameLoadRequest&& request, Event* event, Ref
-        loadPostRequest(WTFMove(request), referrer, loadType, event, WTFMove(formState), WTFMove(completionHandler));
-    else
-        loadURL(WTFMove(request), referrer, loadType, event, WTFMove(formState), WTFMove(privateClickMeasurement), WTFMove(completionHandler));
+    auto finishLoadFrameRequest = [referrer, event, loadType, completionHandler] (Ref<LocalFrame>&& frame, FrameLoadRequest&& request, RefPtr<FormState>&& formState, std::optional<PrivateClickMeasurement>&& privateClickMeasurement) mutable {
```
This looks like it's probably an unnecessary copy of completionHandler.  Could you make completionHandler a CompletionHandler instead of a lambda, then fix it by moving?

rwlbuis replied:
Done.

rwlbuis committed:
[Navigation] Dispatch NavigateEvent on reload
https://bugs.webkit.org/show_bug.cgi?id=273688

Reviewed by Alex Christensen.

Various kinds of reloads, such as navigation.reload() and location.reload() need to dispatch the NavigateEvent [1]. Since a reload triggered
by browser UI (i.e. clicking the reload button on the crome UI) should not dispatch the NavigateEvent, put the logic in FrameLoader::loadFrameRequest
and not in FrameLoader::reload. And for the same reason, make Navigation::reload go through FrameLoader::changeLocation and thus FrameLoader::loadFrameRequest.

This PR also introduces logic to deal with a reload causing an existing window, and matching navigation, to be discarded in favour of a new window and navigation
combination, in that case we have to copy over the navigation entries from the previous Navigation instance.

[1] https://html.spec.whatwg.org/multipage/browsing-the-web.html#reloading-and-traversing:fire-a-push/replace/reload-navigate-event

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/currententrychange-event/navigation-reload-intercept-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/navigate-destination-getState-reload-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-history-entry/key-id-location-reload-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/reload-info-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/reload-no-args-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/reload-state-and-info-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/reload-state-undefined-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/reload-detach-in-onnavigate-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/reload-intercept-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/reload-intercept-rejected-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/reload-preventDefault-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/reload-rejection-order-unload-unserializablestate-expected.txt: Removed.
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/per-entry-events/dispose-same-document-reload-with-intercept-expected.txt:
* Source/WebCore/loader/DocumentWriter.cpp:
(WebCore::DocumentWriter::begin):
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::didBeginDocument):
(WebCore::FrameLoader::loadFrameRequest):
(WebCore::FrameLoader::loadURL):
(WebCore::FrameLoader::loadPostRequest):
(WebCore::FrameLoader::dispatchNavigateEvent):
(WebCore::FrameLoader::updateNavigationAPIEntries): Deleted.
* Source/WebCore/loader/FrameLoader.h:
* Source/WebCore/page/Navigation.cpp:
(WebCore::Navigation::initializeForNewWindow):
(WebCore::Navigation::reload):
(WebCore::Navigation::notifyCommittedToEntry):
(WebCore::Navigation::abortOngoingNavigation):
(WebCore::Navigation::innerDispatchNavigateEvent):
(WebCore::Navigation::initializeEntries): Deleted.
* Source/WebCore/page/Navigation.h:
* Source/WebCore/page/NavigationHistoryEntry.cpp:
(WebCore::NavigationHistoryEntry::NavigationHistoryEntry):
(WebCore::NavigationHistoryEntry::create):
(WebCore::NavigationHistoryEntry::getState const):
(WebCore::NavigationHistoryEntry::setState):
* Source/WebCore/page/NavigationHistoryEntry.h:

Canonical link: https://commits.webkit.org/283604@main
